Revert async formatters

retrohacker opened the pull request:
## Pre-Submission Checklist

- [X] Opened an issue discussing these changes before opening the PR
- [X] Ran the linter and tests via `make prepush`
- [X] Included comprehensive and convincing tests for changes

## Issues

Closes:

* Issue #1347 

> Summarize the issues that discussed these changes

Async formatters were introduced as a way to handle formatting an response body by binding it to data coming from an asyncronous data source. This introduced complexities that were difficult to reconsile with the rest of the code base. By requiring all formatters to be syncronous we simplify the codebase and can :clap: **ship 5.x** :clap: 

# Changes

> What does this PR do?

* Removes asynchronous formatters
* Subjectively simplifies `__send` implementation

retrohacker commented:
Unit tests need to be updated to validate the new behaviour, but would like a review first in case there are any major changes :smile:

rajatkumar commented:
On the surface this looks good but I would like to see the unit tests as unit tests help me with understanding the code better.

hetul commented on lib/response.js:
```diff
@@ -269,13 +209,11 @@ Response.prototype.link = function link(l, rel) {
  * @param    {Object | Buffer | Error} body the content to send
  * @param    {Object} headers  any add'l headers to set
- * @param    {Function} callback a callback for use with async formatters
- * @returns  {Object | undefined} when async formatter in use, returns null,
```
You may have already planned to do this, but consider updating the `response.md` documentation to be consistent with the new JSDoc, just so you don't forget to do it later.

retrohacker replied:
Good catch! I hadn't looked at documentation, would have missed this :heart:

retrohacker commented:
Tests caught a bug where `response.send` and `response.sendRaw` didn't have optional parameters so the argument parser in `__send` was broken. Fixed using `Array.prototype.push`.

Removed all async formatter specific tests.

Only tests failing now are node 8 specific.

Ready for another review, assuming everyone is :+1: on this it's ready for a merge ^_^

hetul commented:
I suppose fixing node 8 support is on another ticket? In any case, LGTM! 👍

rajatkumar commented on lib/response.js:
```diff
@@ -284,114 +224,134 @@ Response.prototype.send = function send(code, body, headers, callback) {
+    // Populate our response object with the derived arguments
+    self.statusCode = code;
+    self._body = body;
+    Object.keys(headers).forEach(function (k) {
```
Good to see the headers are being set here instead of `_flush()`

rajatkumar commented on lib/response.js:
```diff
@@ -404,68 +364,75 @@ function __send(maybeCode, maybeBody, maybeHeaders, maybeCallback, format) {
+        type = mime.lookup(type);
+    }
+
+    if (!self.formatters[type] && self.acceptable.indexOf(type) === -1) {
```
Can we add a code comment explaining why we set type to `application/octet-stream`

retrohacker commented:
Only tests failing on CI are v8 related. Squashing and merging :tada:
